File: LeetCode/April Challenge/20th Day/Day_20.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* constructBinTree(TreeNode *root, int &cval) {
        if(root==NULL) {
            root=new TreeNode(cval);
            
        } else if(cval < root->val) {
            root->left = constructBinTree(root->left, cval);
        } else if( cval> root->val) {
            root->right = constructBinTree(root->right, cval);
        }
        return root;
        
    }
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        TreeNode *root = constructBinTree(root,preorder[0]);
        for(int i = 1; i<preorder.size() ;i++) {
            constructBinTree(root,preorder[i]);
        }
        return root;
    }
    
};

File: LeetCode/April Challenge/20th Day/Day_20.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> TreeNode:
        root = TreeNode(preorder[0])
        
        stack = []
        stack.append(root)
        i = 1
        while i < len(preorder):
            temp = None
            while len(stack) > 0 and preorder[i] > stack[-1].val:
                temp = stack.pop()
            
            if temp != None:
                temp.right = TreeNode(preorder[i])
                stack.append(temp.right)
            
            else:
                temp = stack[-1]
                temp.left = TreeNode(preorder[i])
                stack.append(temp.left)
            i = i + 1
        return root
